### PR TITLE
Remove BN dependency, add ethereumjs-util dependency

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Seems like this library uses "copies" of methods we implemeted in `ethereumjs-util`. This PR removes these copies and instead relies upon `ethereumjs-util`.

There are some type errors which I did not further look into.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -199,65 +199,3 @@ function _decode(input: Buffer): Decoded {
...
-}
-
-/** Transform anything into a Buffer */
-function toBuffer(v: Input): Buffer {
-  if (!Buffer.isBuffer(v)) {
-    if (typeof v === 'string') {
-      if (isHexPrefixed(v)) {
-        return Buffer.from(padToEven(stripHexPrefix(v)), 'hex')
-      } else {
-        return Buffer.from(v)
```
Comment (ryanio) [inline src/index.ts:243]: Hmmm this might unfortunately be a breaking change because this version of toBuffer doesn't have the strictness that our latest util toBuffer has regarding 0x-prefixed strings.

>   1) RLP encoding (string):
       should return itself if single byte and less than 0x7f::
     Error: Cannot convert string to buffer. toBuffer only supports 0x-prefixed hex strings and this string was given: a

Comment (ryanio) [reply src/index.ts:243]: maybe we save this PR for a v3 release, I don't think there is anything else planned for a next breaking change yet though, although one on the backburner that could be good to include would be #95

Comment (ryanio) [reply src/index.ts:243]: might be able to work around this for now by just adding 0x prefix to strings without it in encode, and then disallow it again  in v3 for improved validation, I remember @alcuadrado saying there can be weird bugs when just accepting strings and converting to hex when they don't explicitly have the hex prefix.

**Comment by ryanio** [comment]
another difference here is that util's `intToHex` returns with hex prefixed but the internal lib didn't. We should review the differences in the removed functions carefully as to not cause any accidental breakage. I believe some of the currently failing tests are pointing at these issues.

**Comment by ryanio** [comment]
ok have to take a break but cleaned up some stuff along the way. switched to using ToBufferInputTypes so it can handle future additions. just a few tests remaining failing: integer zero case, longstring2,zero,longList2

haven't looked into these yet, feel free to pick up or I'll continue tomorrow

File: .eslintrc.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ module.exports = {
   },
   env: {
     mocha: true,
+    es2020: true
```
Comment (jochem-brouwer) [inline .eslintrc.js:8]: Added this for eslint BigInt support

File: test/dataTypes.spec.ts [hunk 3]
```diff
@@ -416,3 +436,42 @@ describe('recursive typings', function () {
     assert.ok(assertType<typeof a, Buffer>(true))
   })
 })
+
+describe('getLength', function () {
+  it('should return the right RLP length', function () {
+    assert.ok((<Buffer>RLP.getLength(0)).equals(Buffer.from([])))
```
Comment (jochem-brouwer) [inline test/dataTypes.spec.ts:442]: Was confused while writing this test. Why does `getLength` return empty buffer in case that the input has zero length?

File: test/dataTypes.spec.ts [hunk 4]
```diff
@@ -416,3 +436,42 @@ describe('recursive typings', function () {
...
+    const encodedString = RLP.encode(str)
+    const encodedLongString = RLP.encode(longString)
+    const encodedList = RLP.encode([1, 2, 3])
+    const bufferStream = Buffer.concat([
+      encodedNumber,
+      encodedString,
+      encodedLongString,
+      encodedList,
+    ])
+    let decoded = <RLP.Decoded>(<unknown>RLP.decode(bufferStream, true))
```
Comment (jochem-brouwer) [inline test/dataTypes.spec.ts:464]: Type conversion is necessary for some reason, we should update the function signature.

File: test/dataTypes.spec.ts (hunk 3 above)
Comment (acolytec3) [reply test/dataTypes.spec.ts:456]: If `getLength` is used as part of the decoding process, it seems like it would just be following the definition of rlp encoded integers such that 0 is encoded its most compact form or the empty buffer.  I looked at the line history and don't see any specific commentary on this in past PRs.

Comment (jochem-brouwer) [reply test/dataTypes.spec.ts:456]: You would think it is part of the decoding process, but actually it is not. It is a standalone method and in `index.ts` you can remove it without breaking anything. So it is an external method to "get the length of the RLP input" (what does this mean anyways? That would be just the input Buffer length..? Or do we mean the RLP data length here?), why would this return a number in most cases and a Buffer in edge cases?

Comment (acolytec3) [reply test/dataTypes.spec.ts:456]: In that case, let's fix it.  This is ancient code that's due for a refresh, right?  If we're doing a breaking release, let's rationalize the API and return a 0 here for a 0 length input.

File: src/index.ts [hunk 5]
```diff
@@ -19,12 +18,28 @@ export function encode(input: Input): Buffer {
...
-      : Buffer.concat([encodeLength(inputBuf.length, 128), inputBuf])
   }
+
+  // Edge case, see https://eth.wiki/en/fundamentals/rlp
+  // "The integer 0 = [ 0x80 ]"
+  // EthereumJS-util would convert 0 to the Buffer.from('00', 'hex'), which would
+  // thus return Buffer.from('00', 'hex') when encoding that: this does not match the spec.
+  if (input === 0 || input === BigInt(0)) {
+    return Buffer.from('80', 'hex')
+  }
```
Comment (ryanio) [inline src/index.ts:29]: suggestion to rewrite the comment more succinctly

also adding the BN case I suggested in chat

(might need lint formatting)

could it be an alternative to do this check *after* `toBuffer`? `if (inputBuf.equals(Buffer.from('00', 'hex')) { return Buffer.from('80', 'hex') }`? then we don't have to check the varying input types and can rely on `toBuffer` to handle the input conversion for us and just check if the output matches `Buffer.from('00', 'hex')`

```suggestion
  // According to RLP spec (https://eth.wiki/en/fundamentals/rlp):
  // "The integer 0 = [ 0x80 ]"
  // util's toBuffer converts 0 to Buffer.from('00', 'hex')
  // so here we will return early to match the spec
  if (input === 0 || input === BigInt(0) || (BN.isBN(input) && input.isZero())) {
    return Buffer.from('80', 'hex')
  }
```

File: test/dataTypes.spec.ts (hunk 3 above)
Comment (ryanio) [reply test/dataTypes.spec.ts:456]: thanks @acolytec3 and @jochem-brouwer for opening a PR, will resolve this conversation and can continue in #106

**Comment by paulmillr** [comment]
Strong 👎, 30 lines of code must not be replaced by 300-700KB of dependencies. See #90.

**Comment by paulmillr** [comment]
Can be closed now. cc @ryanio

**Comment by ryanio** [comment]
I'll bring the newly added tests from here to my new PR.
